File: Backend/model/Driver.js
const mongoose = require("mongoose");
const bcrypt = require("bcrypt");

const driverSchema = new mongoose.Schema({
  image: {
    type: String, // URL to image or base64 encoded image
  },
  name: {
    type: String,
    required: true,
    trim: true,
  },
  phone: {
    type: String,
    required: true,
    unique: true,
  },
  email: {
    type: String,
    required: true,
    unique: true,
    lowercase: true,
  },
  password: {
    type: String,
    required: true,
  },
  address: {
    type: String,
    required: true,
  },
  busDetails: {
    busColor: String,
    busModel: String,
    busNumber: {
      type: String,
      required: true,
      unique: true,
    }
  },
  vehicle: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Vehicle'
  }
}, { timestamps: true });

// Password hashing middleware
driverSchema.pre("save", async function () {
  if (!this.isModified('password')) return;
  
  try {
    const salt = await bcrypt.genSalt(10);
    this.password = await bcrypt.hash(this.password, salt);
  } catch (error) {
    throw error;
  }
});

const Driver = mongoose.model("Driver", driverSchema);
module.exports = Driver;

File: Frontend/to_do/lib/models/user/driver.dart
class Driver {
  final String id;
  final String name;
  final String phone;
  final String email;
  final String address;
  final BusDetails busDetails;
  final String? licenseNumber;
  final String? image;

  Driver({
    required this.id,
    required this.name,
    required this.phone,
    required this.email,
    required this.address,
    required this.busDetails,
    this.licenseNumber,
    this.image,
  });

  factory Driver.fromJson(Map<String, dynamic> json) {
    return Driver(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      phone: json['phone'] ?? '',
      email: json['email'] ?? '',
      address: json['address'] ?? '',
      busDetails: BusDetails.fromJson(json['busDetails'] ?? {}),
      licenseNumber: json['licenseNumber'],
      image: json['image'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'phone': phone,
      'email': email,
      'address': address,
      'busDetails': busDetails.toJson(),
      'licenseNumber': licenseNumber,
      'image': image,
    };
  }
}

class BusDetails {
  final String? busColor;
  final String? busModel;
  final String busNumber;

  BusDetails({
    this.busColor,
    this.busModel,
    required this.busNumber,
  });

  factory BusDetails.fromJson(Map<String, dynamic> json) {
    return BusDetails(
      busColor: json['busColor'],
      busModel: json['busModel'],
      busNumber: json['busNumber'] ?? 'Unknown',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'busColor': busColor,
      'busModel': busModel,
      'busNumber': busNumber,
    };
  }
}
